src/alexapi/playback_handlers/basehandler.py

```diff
@@ -1,41 +1,171 @@
 from abc import ABCMeta, abstractmethod
+import threading
+import time
+from collections import deque
+
+import alexapi.bcolors as bcolors
+from alexapi.constants import RequestType, PlayerActivity
+
+
+class PlaybackAudioType:
+	SPEECH = 'speech'
+	MEDIA = 'media'
+
+
+class PlaybackItem:
+	def __init__(self, url, offset, audio_type, stream_id):
+		self.url = url
+		self.offset = offset
+		self.audio_type = audio_type
+		self.stream_id = stream_id
+
+
+class PlaybackLock:
+	def __init__(self):
+		# This has inverted logic
+		self.play_lock = threading.Event()
+		self.play_lock.set()
+
+	def acquire(self):
+		if not self.play_lock.isSet():
+			self.play_lock.wait()
+		self.play_lock.clear()
+		self.is_playing = True
+
+	def release(self):
+		self.play_lock.set()
+		self.is_playing = False
 
 
 class BaseHandler:
 	__metaclass__ = ABCMeta
 
-	is_playing = False
-	volume = None
-	media_volume = None
+	def __init__(self, config, callback_report):
+		self.__config = config
+		self.__callback_report = callback_report
 
-	@abstractmethod
-	def __init__(self, config):
-		pass
+		self.queue = None
+		self.processing_queue = False
+
+		self.queue = deque()
+		self.processing_queue = False
+
+		self.stream_id = None
+		self.is_playing = False
+
+		self.play_lock = PlaybackLock()
 
 	@abstractmethod
-	def setup(self):
+	def on_setup(self):
 		pass
 
 	@abstractmethod
-	def queued_play(self, url, offset=0, audio_type='media', streamId=None):
+	def on_play(self, item):
 		pass
 
 	@abstractmethod
-	def blocking_play(self, url, offset=0, audio_type='speech', streamId=None):
+	def on_stop(self):
 		pass
 
 	@abstractmethod
-	def stop(self):
+	def on_cleanup(self):
 		pass
 
 	@abstractmethod
-	def set_volume(self, volume):
+	def on_set_volume(self, volume):
 		pass
 
 	@abstractmethod
-	def set_media_volume(self, volume):
+	def on_set_media_volume(self, volume):
 		pass
 
-	@abstractmethod
+	def report_play(self, debug_msg=None):
```

Comment:
The report_ methods are superior to depending on each subclass to report the states and activity correctly. Encapsulation behind the report_ methods makes playback handler implementations simpler. This prevents errors like those identified in #117